### PR TITLE
[PLUGIN-1757] Remove Page iteration from transform

### PR DESCRIPTION
**Pull request by psainics**
##  Remove Page iteration from transform

Jira : [PLUGIN-1757](https://cdap.atlassian.net/browse/PLUGIN-1757)

### Description

Plugin had very high memory usage compared to the input file size.

This PR improves memory usage by removing Page iteration from transform.

### OLD
image

### NEW
image

### UI Field

- No Changes made to widget json.

### Docs

- No Changes made to docs.

### Code change

- Modified `HttpBatchSource.java`
- Modified `HttpInputFormat.java`
- Modified `HttpRecordReader.java`

### Unit Tests

- No Changes made to unit tests.

### Results

- This PR ( Before removing page iteration and current implementation )
image

- This PR with closing the basePage
image

### CODE REVIEW

**Comment by MrRahulSharma** [comment]
Can you add the testing details performed for this change ?

**Comment by psainics** [comment]
> Can you add the testing details performed for this change ?

For performance i have tested this with a 2GB csv file.
We saw memory usage reduced from 7GB to 1.5GB.

I have done pagination test to make sure it's same.

Testing for other formats is in being done !

**Review by MrRahulSharma — APPROVED** [review]

File: src/main/java/io/cdap/plugin/http/source/batch/HttpRecordReader.java [hunk 1]
```diff
@@ -57,10 +59,22 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
...
-      return false;
+    // If there is no current page or no next line in the current page
+    if (currentBasePage == null || !currentBasePage.hasNext()) {
+      if (!httpPaginationIterator.hasNext()) {
+        // If there is no next page, return false
+        // All pages are read
+        return false;
+      }
+      // Get the next page
+      currentBasePage = httpPaginationIterator.next();
```
Comment (arjan-bal) [inline src/main/java/io/cdap/plugin/http/source/batch/HttpRecordReader.java:70]: We are not closing the currentBasePage here. This [can hold a BufferedInputReader](https://github.com/data-integrations/http/blob/557bea0b1dd31351b171572e13e428dd9bbd8cc5/src/main/java/io/cdap/plugin/http/common/pagination/page/RecordPerLinePage.java#L85-L89). If we don't close it, its memory may still be freed when it is finalized.

Can you try closing the page to see if it improves the memory utilization further?

Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/http/source/batch/HttpRecordReader.java:70]: We have looked into this and closing the basepage is not making any impact on the memory usage, but its better to close it and we will do the same in next release. Adding snapshot in the description for memory usage after closing the reader.

Comment (MrRahulSharma) [reply src/main/java/io/cdap/plugin/http/source/batch/HttpRecordReader.java:70]: Good catch. We should make sure that we do close it.

**Pull request merged**